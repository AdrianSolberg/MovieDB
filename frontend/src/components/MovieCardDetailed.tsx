import React from "react";
import {
    Card,
    CardContent,
    CardHeader,
    CardTitle,
} from "../shadcn/components/ui/card";
import { Button } from "../shadcn/components/ui/button";
import Ratings from "../shadcn/components/ui/rating";
import { getImageUrl, ImageType } from "@/utils/imageUrl/imageUrl";
import { Movie } from "@/types/__generated__/types";

interface MovieCardDetailedProps {
    movie: Movie;
}

const formatCurrency = (value: number) => {
    return value.toLocaleString("en-US", {
        style: "currency",
        currency: "USD",
        maximumFractionDigits: 0,
    });
};

const getMovieDetailsList = (movie: Movie) => [
    {
        label: "Status",
        value: movie.status,
    },
    {
        label: "Budget",
        value: movie.budget ? `${formatCurrency(movie.budget)}` : "N/A",
    },
    {
        label: "Revenue",
        value: movie.revenue ? `${formatCurrency(movie.revenue)}` : "N/A",
    },
    {
        label: "Production Companies",
        value: movie.production_companies.length
            ? movie.production_companies.join(", ")
            : "N/A",
    },
    {
        label: "Production Countries",
        value: movie.production_countries.length
            ? movie.production_countries.join(", ")
            : "N/A",
    },
    {
        label: "Spoken Languages",
        value: movie.spoken_languages.length
            ? movie.spoken_languages.join(", ")
            : "N/A",
    },
];

const MovieCardDetailed: React.FC<MovieCardDetailedProps> = ({ movie }) => {
    const releaseYear = movie.release_date.getFullYear();
    const runtimeHours = Math.floor(movie.runtime / 60);
    const runtimeMinutes = movie.runtime % 60;

    return (
        <Card
            className="relative m-4 overflow-hidden bg-cover bg-center shadow-lg"
            style={{
                backgroundImage: `url(${getImageUrl(ImageType.BACKDROP, movie.backdrop_path)})`,
            }}
            data-testid="detailedMovieCard"
        >
            {movie.backdrop_path && (
                <div className="absolute inset-0 bg-black opacity-85" />
            )}
            <CardContent className="relative h-full p-6 text-white">
                <CardHeader className="mb-4 p-0">
                    <CardTitle className="text-5xl font-bold">
                        {movie.title}
                    </CardTitle>
                    <p className="text-sm">
                        {[
                            releaseYear,
                            `${runtimeHours}h ${runtimeMinutes}m`,
                            movie.genres.length
                                ? movie.genres.join(", ")
                                : null,
                        ]
                            .filter(Boolean)
                            .join(" â€¢ ")}
                    </p>
                    <Ratings
                        value={movie.vote_average / 2}
                        size={24}
                        variant="yellow"
                        totalstars={5}
                        title={String((movie.vote_average / 2).toFixed(2))}
                    />
                </CardHeader>
                <section className="flex h-full flex-col md:flex-row">
                    <figure className="mb-6 w-full max-w-72 flex-shrink-0 md:mb-0 md:mr-6">
                        <img
                            src={getImageUrl(
                                ImageType.POSTER,
                                movie.poster_path,
                                "w500"
                            )}
                            className="rounded-lg"
                            alt={`Poster of ${movie.title}`}
                            title={movie.title}
                        />
                    </figure>
                    <section className="flex-grow">
                        {movie.tagline && (
                            <p className="mb-2 text-xl italic text-gray-300">
                                {movie.tagline}
                            </p>
                        )}
                        <p className="mb-4 text-lg">{movie.overview}</p>
                        <ul className="mb-6 flex flex-col gap-2">
                            {getMovieDetailsList(movie).map((item, index) => (
                                <li key={index}>
                                    <strong>{item.label}: </strong>
                                    {item.value}
                                </li>
                            ))}
                        </ul>
                        {movie.homepage && (
                            <Button asChild>
                                <a
                                    href={movie.homepage}
                                    target="_blank"
                                    rel="noopener noreferrer"
                                >
                                    Visit Homepage
                                </a>
                            </Button>
                        )}
                    </section>
                </section>
            </CardContent>
        </Card>
    );
};

export default MovieCardDetailed;
