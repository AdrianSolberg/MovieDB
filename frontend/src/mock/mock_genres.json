[
    {
        "id": 0,
        "name": "Drama"
    },
    {
        "id": 1,
        "name": "Romance"
    },
    {
        "id": 2,
        "name": "Western"
    },
    {
        "id": 3,
        "name": "Animation"
    },
    {
        "id": 4,
        "name": "War"
    },
    {
        "id": 5,
        "name": "Science Fiction"
    },
    {
        "id": 6,
        "name": "Comedy"
    },
    {
        "id": 7,
        "name": "Fantasy"
    },
    {
        "id": 8,
        "name": "Adventure"
    },
    {
        "id": 9,
        "name": "Crime"
    },
    {
        "id": 10,
        "name": "Thriller"
    },
    {
        "id": 11,
        "name": "Family"
    },
    {
        "id": 12,
        "name": "Action"
    },
    {
        "id": 13,
        "name": "Mystery"
    }
]
