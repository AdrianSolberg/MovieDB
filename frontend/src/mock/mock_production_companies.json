[
    "Lynda Obst Productions",
    "Pascal Pictures",
    "Fairview Entertainment",
    "Silver Pictures",
    "Walt Disney Pictures",
    "Visiona Romantica",
    "DC Films",
    "Mandate International",
    "Patalex IV Productions Limited",
    "Amblin Entertainment",
    "Heyday Films",
    "Universal Pictures",
    "Bron Studios",
    "Juno Pix",
    "Color Force",
    "LStar Capital",
    "Miramax",
    "The Saul Zaentz Company",
    "Sikelia Productions",
    "Red Granite Pictures",
    "The Weinstein Company",
    "Knickerbocker Films",
    "Columbia Pictures",
    "Groucho II Film Partnership",
    "Syncopy",
    "P of A Productions Limited",
    "Zehnte Babelsberg Film",
    "The Linson Company",
    "Pixar",
    "Appian Way",
    "The Steve Tisch Company",
    "Genre Films",
    "Paramount",
    "Kennedy Miller Productions",
    "Lucasfilm Ltd.",
    "New Line Cinema",
    "Atman Entertainment",
    "Cecchi Gori Pictures",
    "Castle Rock Entertainment",
    "Isobel Griffiths",
    "Phoenix Pictures",
    "DC Entertainment",
    "Dune Entertainment",
    "20th Century Fox",
    "Marvel Studios",
    "Lightstorm Entertainment",
    "WingNut Films",
    "Fox 2000 Pictures",
    "RatPac Entertainment",
    "Lionsgate",
    "Warner Bros. Pictures",
    "Patalex III Productions Limited",
    "Jersey Films",
    "Joint Effort",
    "Village Roadshow Pictures",
    "EMJAG Productions",
    "The Donners' Company",
    "1492 Pictures",
    "A Band Apart",
    "DC Comics",
    "Atlas Entertainment",
    "Ingenious Media",
    "Legendary Pictures",
    "Taurus Film",
    "Wendy Finerman Productions",
    "Peninsula Films",
    "Lin Pictures",
    "Regency Enterprises",
    "Jerry Bruckheimer Films"
]
