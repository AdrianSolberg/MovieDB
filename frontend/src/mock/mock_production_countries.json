[
    {
        "id": 0,
        "name": "New Zealand"
    },
    {
        "id": 1,
        "name": "Canada"
    },
    {
        "id": 2,
        "name": "South Africa"
    },
    {
        "id": 3,
        "name": "United Kingdom"
    },
    {
        "id": 4,
        "name": "Germany"
    },
    {
        "id": 5,
        "name": "Australia"
    },
    {
        "id": 6,
        "name": "France"
    },
    {
        "id": 7,
        "name": "United States of America"
    }
]
