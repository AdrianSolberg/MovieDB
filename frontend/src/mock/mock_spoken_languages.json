[
    {
        "id": 0,
        "name": "Swahili"
    },
    {
        "id": 1,
        "name": "French"
    },
    {
        "id": 2,
        "name": "German"
    },
    {
        "id": 3,
        "name": "Persian"
    },
    {
        "id": 4,
        "name": "Italian"
    },
    {
        "id": 5,
        "name": "Japanese"
    },
    {
        "id": 6,
        "name": "Korean"
    },
    {
        "id": 7,
        "name": "Hindi"
    },
    {
        "id": 8,
        "name": "Urdu"
    },
    {
        "id": 9,
        "name": "English"
    },
    {
        "id": 10,
        "name": "Mandarin"
    },
    {
        "id": 11,
        "name": "Spanish"
    },
    {
        "id": 12,
        "name": "Russian"
    },
    {
        "id": 13,
        "name": "Xhosa"
    },
    {
        "id": 14,
        "name": "Romanian"
    },
    {
        "id": 15,
        "name": "Norwegian"
    },
    {
        "id": 16,
        "name": "Swedish"
    },
    {
        "id": 17,
        "name": "Arabic"
    }
]
