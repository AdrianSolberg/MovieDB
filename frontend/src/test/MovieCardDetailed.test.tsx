import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import MovieCardDetailed from "../components/MovieCardDetailed";
import { all_movies } from "./mock/util";
import { vi } from "vitest";

vi.mock("@/utils/formatUtil", () => ({
    formatCurrency: (value: number) =>
        value.toLocaleString("en-US", {
            style: "currency",
            currency: "USD",
            maximumFractionDigits: 0,
        }),
}));

vi.mock("@/utils/imageUrl/imageUrl", () => ({
    ImageType: {
        POSTER: "poster",
        BACKDROP: "backdrop",
    },
    getImageUrl: vi.fn((type, path, size) => {
        return type === "poster"
            ? `mocked_poster_url/${size}${path}`
            : `mocked_backdrop_url/original${path}`;
    }),
}));

const mockMovie = all_movies[18]; // Joker

describe("MovieCardDetailed", () => {
    it("displays movie title, release year, runtime and genres", () => {
        render(<MovieCardDetailed movie={mockMovie} />);
        expect(screen.getByText("Joker")).toBeInTheDocument();
        expect(
            screen.getByText("2019 • 2h 2m • Crime, Thriller, Drama")
        ).toBeInTheDocument();
    });

    it("does not display genres if there are none", () => {
        render(
            <MovieCardDetailed
                movie={{
                    ...mockMovie,
                    genres: [],
                }}
            />
        );
        expect(screen.getByText("2019 • 2h 2m")).toBeInTheDocument();
    });

    it("displays movie rating", () => {
        render(<MovieCardDetailed movie={mockMovie} />);
        expect(screen.getByTitle("4.08")).toBeInTheDocument();
    });

    it("displays movie poster", () => {
        render(<MovieCardDetailed movie={mockMovie} />);

        const poster = screen.getByAltText("Poster of Joker");
        expect(poster).toBeInTheDocument();

        expect(poster).toHaveAttribute(
            "src",
            "mocked_poster_url/w500/udDclJoHjfjb8Ekgsd4FDteOkCU.jpg"
        );
    });

    it("displays movie tagline and overview", () => {
        render(<MovieCardDetailed movie={mockMovie} />);
        expect(screen.getByText("Put on a happy face.")).toBeInTheDocument();
        expect(screen.getByText(mockMovie.overview)).toBeInTheDocument();
    });

    it("does not display movie tagline if it is missing", () => {
        render(
            <MovieCardDetailed
                movie={{
                    ...mockMovie,
                    tagline: null,
                }}
            />
        );
        expect(
            screen.queryByText("Put on a happy face.")
        ).not.toBeInTheDocument();
    });

    it("displays movie details", () => {
        render(<MovieCardDetailed movie={mockMovie} />);
        expect(screen.getByText("Released")).toBeInTheDocument();
        expect(screen.getByText("$55,000,000")).toBeInTheDocument();
        expect(screen.getByText("$1,074,458,282")).toBeInTheDocument();
        expect(
            screen.getByText(
                "Warner Bros. Pictures, Joint Effort, Village Roadshow Pictures, Bron Studios, DC Films"
            )
        ).toBeInTheDocument();
        expect(
            screen.getByText("Canada, United States of America")
        ).toBeInTheDocument();
        expect(screen.getByText("English")).toBeInTheDocument();
    });

    it("displays N/A for missing fields", () => {
        render(
            <MovieCardDetailed
                movie={{
                    ...mockMovie,
                    budget: 0,
                    revenue: 0,
                    production_companies: [],
                    production_countries: [],
                    spoken_languages: [],
                }}
            />
        );
        expect(screen.getAllByText("N/A")).toHaveLength(5);
    });

    it("displays 'Visit Homepage' button with link", () => {
        render(<MovieCardDetailed movie={mockMovie} />);
        const homepageButton = screen.getByRole("link", {
            name: "Visit Homepage",
        });
        expect(homepageButton).toBeInTheDocument();
        expect(homepageButton).toHaveAttribute("href", mockMovie.homepage);
    });

    it("does not display 'Visit Homepage' button if homepage is missing", () => {
        render(
            <MovieCardDetailed
                movie={{
                    ...mockMovie,
                    homepage: null,
                }}
            />
        );
        const homepageButton = screen.queryByRole("link", {
            name: "Visit Homepage",
        });
        expect(homepageButton).not.toBeInTheDocument();
    });

    it("renders background image correctly", () => {
        render(<MovieCardDetailed movie={mockMovie} />);

        expect(screen.getByTestId("detailedMovieCard")).toHaveStyle(
            `background-image: url(mocked_backdrop_url/original/hO7KbdvGOtDdeg0W4Y5nKEHeDDh.jpg)`
        );
    });
});
