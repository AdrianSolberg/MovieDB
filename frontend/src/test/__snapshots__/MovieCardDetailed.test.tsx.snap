// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MovieCardDetailed > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="rounded-lg border bg-card text-card-foreground relative m-4 overflow-hidden bg-cover bg-center shadow-lg"
    data-testid="detailedMovieCard"
    style="background-image: url(mocked_backdrop_url/original/hO7KbdvGOtDdeg0W4Y5nKEHeDDh.jpg);"
  >
    <div
      class="absolute inset-0 bg-black opacity-85"
    />
    <div
      class="relative h-full p-6 text-white"
    >
      <div
        class="flex flex-col space-y-1.5 mb-4 p-0"
      >
        <h1
          class="tracking-tight text-5xl font-bold"
        >
          Joker
        </h1>
        <p
          class="text-sm"
        >
          <time
            datetime="2019"
          >
            2019
          </time>
           • 
          <time
            datetime="PT2H2M"
          >
            2h 2m
          </time>
           • 
          <span>
            Crime, Thriller, Drama
          </span>
        </p>
        <div
          class="flex items-center gap-2"
          title="4.08"
        >
          <svg
            class="lucide lucide-star fill-current text-yellow-500"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            tabindex="-1"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <polygon
              points="12 2 15.09 8.26 22 9.27 17 14.14 18.18 21.02 12 17.77 5.82 21.02 7 14.14 2 9.27 8.91 8.26 12 2"
            />
          </svg>
          <svg
            class="lucide lucide-star fill-current text-yellow-500"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            tabindex="-1"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <polygon
              points="12 2 15.09 8.26 22 9.27 17 14.14 18.18 21.02 12 17.77 5.82 21.02 7 14.14 2 9.27 8.91 8.26 12 2"
            />
          </svg>
          <svg
            class="lucide lucide-star fill-current text-yellow-500"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            tabindex="-1"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <polygon
              points="12 2 15.09 8.26 22 9.27 17 14.14 18.18 21.02 12 17.77 5.82 21.02 7 14.14 2 9.27 8.91 8.26 12 2"
            />
          </svg>
          <svg
            class="lucide lucide-star fill-current text-yellow-500"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            tabindex="-1"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <polygon
              points="12 2 15.09 8.26 22 9.27 17 14.14 18.18 21.02 12 17.77 5.82 21.02 7 14.14 2 9.27 8.91 8.26 12 2"
            />
          </svg>
          <div
            class="relative inline-block"
          >
            <svg
              class="lucide lucide-star fill-transparent text-yellow-500"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <polygon
                points="12 2 15.09 8.26 22 9.27 17 14.14 18.18 21.02 12 17.77 5.82 21.02 7 14.14 2 9.27 8.91 8.26 12 2"
              />
            </svg>
            <div
              style="position: absolute; top: 0px; overflow: hidden; width: 8.399999999999963%;"
            >
              <svg
                class="lucide lucide-star fill-current text-yellow-500"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <polygon
                  points="12 2 15.09 8.26 22 9.27 17 14.14 18.18 21.02 12 17.77 5.82 21.02 7 14.14 2 9.27 8.91 8.26 12 2"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
      <section
        class="flex h-full flex-col md:flex-row"
      >
        <figure
          class="mb-6 w-full max-w-72 flex-shrink-0 md:mb-0 md:mr-6"
        >
          <img
            alt="Poster of Joker"
            class="rounded-lg"
            src="mocked_poster_url/w500/udDclJoHjfjb8Ekgsd4FDteOkCU.jpg"
            title="Joker"
          />
        </figure>
        <section
          class="flex-grow"
        >
          <p
            class="mb-2 text-xl italic text-gray-300"
          >
            Put on a happy face.
          </p>
          <p
            class="mb-4 text-lg"
          >
            During the 1980s, a failed stand-up comedian is driven insane and turns to a life of crime and chaos in Gotham City while becoming an infamous psychopathic crime figure.
          </p>
          <ul
            class="mb-6 flex flex-col gap-2"
          >
            <li>
              <strong>
                Status: 
              </strong>
              Released
            </li>
            <li>
              <strong>
                Budget: 
              </strong>
              $55,000,000
            </li>
            <li>
              <strong>
                Revenue: 
              </strong>
              $1,074,458,282
            </li>
            <li>
              <strong>
                Production Companies: 
              </strong>
              Warner Bros. Pictures, Joint Effort, Village Roadshow Pictures, Bron Studios, DC Films
            </li>
            <li>
              <strong>
                Production Countries: 
              </strong>
              Canada, United States of America
            </li>
            <li>
              <strong>
                Spoken Languages: 
              </strong>
              English
            </li>
          </ul>
          <a
            class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground hover:bg-primary/90 h-10 px-4 py-2"
            href="http://www.jokermovie.net/"
            rel="noopener noreferrer"
            target="_blank"
          >
            Visit Homepage
          </a>
        </section>
      </section>
    </div>
  </div>
</DocumentFragment>
`;
